### ISSUE TITLE
[google_maps_flutter] Performance drop when loading markers from bytes

### ISSUE THREAD

**Issue by theochampion** [flutter/flutter#41731]
Hi there,

I'm building a map application in which I have A LOT of different marker designs to display (depending on the kind of POI).
As I have to constantly renderer different marker on the map, I decided to optimize performance by loading all the marker design from assets file (using `BitmapDescriptor.fromImageAsset()` function) at the beginning and storing that in a
```dart
Map<int, BitmapDescriptor>
```
It worked great, the performance was good.

I then decided to generate my markers with code instead of loading them from image assets using the following code:

```dart
Future<BitmapDescriptor> generatePostClusterMarker(
    String text, double sizeFactor, double pixelRatio) async {
  print(pixelRatio);
  Function wRatio = (input) => input * pixelRatio;
  final stopwatch = Stopwatch()..start();
  final ui.PictureRecorder recorder = ui.PictureRecorder();
  final Canvas canvas = Canvas(recorder);

  final TextSpan textSpan = TextSpan(
    style: TextStyle(
        color: Colors.pink,
        fontSize: wRatio(13 * sizeFactor),
        fontWeight: FontWeight.w900),
    text: text,
  );

  final TextPainter textPainter = TextPainter(
      text: textSpan,
      textDirection: TextDirection.ltr,
      textAlign: TextAlign.center);
  textPainter.layout();

  double diameter = 100 * sizeFactor;
  double blurSigma = wRatio(5);
  Size boxSize = Size(diameter, diameter);
  Size outputSize =
      Size(boxSize.width + 2 * blurSigma, boxSize.height + 2 * blurSigma);

  // paint shadow
  canvas.drawCircle(
      Offset(outputSize.width / 2, outputSize.height / 2),
      diameter / 2,
      Paint()
        ..color = Colors.black.withAlpha(50)
        ..maskFilter = MaskFilter.blur(BlurStyle.normal, blurSigma / 2));
  // paint rounded box
  canvas.drawCircle(
      Offset(outputSize.width / 2, outputSize.height / 2),
      diameter / 2,
      Paint()
        ..color = Colors.white
        ..isAntiAlias = true
        ..style = PaintingStyle.fill
        ..filterQuality = FilterQuality.high);
  // paint text
  textPainter.paint(
      canvas,
      Offset((outputSize.width / 2) - (textPainter.size.width / 2),
          (outputSize.height / 2) - (textPainter.size.height / 2)));

  final ui.Picture picture = recorder.endRecording();

  final ByteData pngBytes = await picture
      .toImage(outputSize.width.toInt(), outputSize.height.toInt())
      .then((image) => image.toByteData(format: ui.ImageByteFormat.png));
  print(
      'Marker $text [${outputSize.width}x${outputSize.height}] generated in ${stopwatch.elapsed.inMilliseconds}');

  picture.dispose();
  return BitmapDescriptor.fromBytes(pngBytes.buffer.asUint8List());
}
```

again loading them all at the beginning and storing all the resulting `BitmapDescriptor` in a Map.

But after doing that, the performance took a big hit, markers now take longer to render on the map and the whole thing is not fluid as it used to be when they were loaded from image assets.

What I don't understand is that after the initial loading, the generated `Map<int, BitmapDescriptor>` should contain exactly the same thing (as the generated markers are exported with the same image size as the image assets), so why is performance worst?

What really is the difference between this:
![Screen Shot 2019-10-01 at 8 13 47 pm](https://user-images.githubusercontent.com/14275989/65989694-7325fd00-e48a-11e9-9f90-317b8c7c82d5.png)

and this:

![Screen Shot 2019-10-01 at 8 12 11 pm](https://user-images.githubusercontent.com/14275989/65989731-82a54600-e48a-11e9-8862-83d9851eed62.png)

Am I missing something here? What else can I do to improve marker loading performance? 

Cheers!

Theo

**Comment by chinmaygarde** [comment]
There are plenty of extremely expensive operations Flutter needs to perform to take a compressed (as PNG) image buffer and display it on-screen. The image needs to be decompressed and then uploaded to the GPU (on separate worker threads no less). Depending on whether you are doing those operations per frame, I can see how this would not be performant.

To make sure that you are doing all the work upfront and saving it between frames, try making the map value contain an [Image](https://api.flutter.dev/flutter/dart-ui/Image-class.html) (or something that holds one).

Loading images from assets already does this for you which is probably why this was faster.

**Comment by theochampion** [comment]
@chinmaygarde Thanks for the quick reply!

So if I understand well, I would store my computed markers in a 
```dart
Map<int, Image>
```
instead of a 
```dart
Map<int, BitmapDescriptor>
```
But, as [BitmapDescriptor](https://pub.dev/documentation/google_maps_flutter/latest/google_maps_flutter/BitmapDescriptor-class.html) can only be constructed using byteData (`Uint8List`), I would later have to call the `toByteData` method of `Image` every time I want to create a `BitmapDescriptor`. Isn't that worst than just storing the BitmapDescriptor in the first place?

> There are plenty of extremely expensive operations Flutter needs to perform to take a compressed (as PNG) image buffer and display it on-screen.

Aren't those operations run on the platform side (as they are part of the native maps) ?

Cheers :)

**Comment by chinmaygarde** [comment]
> Isn't that worst than just storing the BitmapDescriptor in the first place? Aren't those operations run on the platform side (as they are part of the native maps) ?

You're right. I was under the impression that the markers were being rendered by Flutter (using a platform view overlay). But [reading the implementation of the plugin](https://github.com/flutter/plugins/blob/master/packages/google_maps_flutter/ios/Classes/), that doesn't seem to be the case. The implementation for markers from assets vs bitmaps seems similar so I am not sure why there is a performance degradation. Can you provide an [Observatory Timeline trace of the scene](https://medium.com/flutter/profiling-flutter-applications-using-the-timeline-a1a434964af3) in Profile mode to this issue please?

**Comment by RyanRamchandar** [comment]
I was attempting to do the same but ran into performance issues (not to mention trying to make everything display density independent) as well trying to create images and load them as an asset. Here's how I was creating the image: https://github.com/brendan-duncan/image/issues/146 I was able to draw a simple circle image in under 10ms.

**Comment by theochampion** [comment]
@chinmaygarde 
Attached you will find two timeline trace of doing the exact same action on the map with BitmapDescriptor loaded from assets (`trace_assets.json`) and BitmapDescriptor loaded from bytes (`trace_bitmap.json`)
[Archive 2.zip](https://github.com/flutter/flutter/files/3678693/Archive.2.zip)

**Comment by theochampion** [comment]
@RyanRamchandar Did you manage to store the precomputed images before using them?

**Comment by RyanRamchandar** [comment]
@theochampion I didn't get that far, unfortunately. But with the Image package it's relatively straightforward to save the generated image to a PNG to other file type.

eg.
```
// Create an image
Image image = Image(101, 101); // circle is imperfect, increase by 1 to fit

for(int x=0; x<=50; x++) {
    drawCircle(image, 50, 50, x, getColor(0,0,255));
}

// Save the image to disk as a PNG
File('test.png').writeAsBytesSync(encodePng(image));
```

**Comment by escamoteur** [comment]
@theochampion what you could also try is to store a List of ByteData and see if that makes a de.

**Comment by theochampion** [comment]
> I didn't get that far, unfortunately, but with the Image package it's relatively straightforward to save the byte stream to a PNG or other file type.

@RyanRamchandar  So you generate actual images on disk and then load them as assets?

> what you could also try is to store a List of ByteData and see if that makes a de.

@escamoteur Yes, this was the first thing I tried, no differences :/

**Comment by escamoteur** [comment]
Perhaps in here might lye an answer https://m.youtube.com/watch?v=r-TLSBdHe1A

**Comment by theochampion** [comment]
@chinmaygarde Did you had chance to look at the profiling trace? 🕵

**Comment by theochampion** [comment]
> Perhaps in here might lye an answer https://m.youtube.com/watch?v=r-TLSBdHe1A

@escamoteur Great presentation, but not sure I understand what you mean 🤔

**Comment by escamoteur** [comment]
My point is, that we have no real explanation why one way is slower.
But for sure you get a different memory layout if you create all images at the startup and keep that in memory compared to load them on demand into the image cache. 
You could give it a try and create a large UInt8List in the faster version at the place where you create your images in the slower version and compare if it is still faster.

**Comment by theochampion** [comment]
I went through the source code of the plugin and it appears that they serialize the `Markers` to send them to the platform view 

google_maps.dart `l. 206`
```dart
    final Map<String, dynamic> creationParams = <String, dynamic>{
      'initialCameraPosition': widget.initialCameraPosition?.toMap(),
      'options': _googleMapOptions.toMap(),
      'markersToAdd': _serializeMarkerSet(widget.markers),
      'polygonsToAdd': _serializePolygonSet(widget.polygons),
      'polylinesToAdd': _serializePolylineSet(widget.polylines),
      'circlesToAdd': _serializeCircleSet(widget.circles),
    };
    if (defaultTargetPlatform == TargetPlatform.android) {
      return AndroidView(
        viewType: 'plugins.flutter.io/google_maps',
        onPlatformViewCreated: onPlatformViewCreated,
        gestureRecognizers: widget.gestureRecognizers,
        creationParams: creationParams,
        creationParamsCodec: const StandardMessageCodec(),
      );
    } 
```

This could lead to a difference in performance in transfer speed because `BitmapDescriptors` created from assets only contain a reference to the file to create the asset from while `BitmapDescriptors` created from bytes contain the actual `Uint8List` data of the image.

Moreover, it appears that if multiple markers have the same `BitmapDescriptor`, the data for that `BitmapDescriptor` is transferred multiple times which can lead to even higher transfer size to the platform view (especially if you have a lot of markers on screen). That would explain why the performance hit grow proportionally to the number of markers onscreen. 

I didn't manage to size or time any of this, but this looks like a big bottleneck.

What do you guys think? @chinmaygarde @escamoteur

**Comment by escamoteur** [comment]
Ah, you mean they only pass down the asset address? That would make totally sense. That they send it multiple times though is bad.
Great research.

**Comment by RyanRamchandar** [comment]
> So you generate actual images on disk and then load them as assets?

No, I was just loading the bytes directly into `BitmapDescriptor.fromBytes`

**Comment by escamoteur** [comment]
@RyanRamchandar but if you do use `fromBytes` then you should have the same problem as @theochampion maybe your markers were smaller?

**Comment by RyanRamchandar** [comment]
@escamoteur yes I had similar performance issues as @theochampion

**Comment by escamoteur** [comment]
I'm sure the Flutter team will look at this when they check all map issues.

**Comment by vtimsit** [comment]
Any update on this ? Having the same problem and it's a big bottleneck

**Comment by escamoteur** [comment]
@chinmaygarde do we have any new insights into this?

**Comment by JBBx2016** [comment]
In order to prevent entire markers to be transmitted to platform on every mutation of the marker, I modified marker_updates.dart to only send the properties changed. This is especially useful when wanting to animate the movements of several 100 markers on the screen.

https://github.com/JBBx2016/plugins/commit/e3db59075f9e4e88ca938cbbe9ab9c93035c62e3

**Comment by RyanRamchandar** [comment]
@JBBx2016 have you opened a PR for this enhancement? That would be a great contribution.

**Comment by theochampion** [comment]
@JBBx2016 So it's an extension of this PR https://github.com/flutter/plugins/pull/1933/commits ?

Happy to help if you need to test / PR :)

**Comment by theochampion** [comment]
I opened a PR to avoid unnecessary updates, this is not a final solution for this as it doesn't deal with the cost of updates themselves but can help for now https://github.com/flutter/plugins/pull/2235

Feel free to upvote it ;)

**Comment by theochampion** [comment]
Hey guys, so I continued my investigation on this and finally manage to narrow down the problem:

So by timing the time taken on the platform side to update the markers like so (in `google_maps_flutter/controller.dart` L.110)

```dart
  Future<void> _updateMarkers(_MarkerUpdates markerUpdates) async {
    final stopwatch = Stopwatch()..start();
    Map<String, dynamic> updates = markerUpdates._toMap();
    print(updates);
    assert(markerUpdates != null);
    await channel.invokeMethod<void>(
      'markers#update',
      updates,
    );
    print(
        "UPDATING (${updates["markersToAdd"].length} add, ${updates["markersToChange"].length} change, ${updates["markerIdsToRemove"].length} delete) MARKERS ON PLATFORM CHANNEL TOOK ${stopwatch.elapsedMilliseconds}ms");
  }
```

I got the following results for the exact same experiment:

With markers loaded from assets:
```
UPDATING (97 add, 0 change, 2 delete) MARKERS ON PLATFORM CHANNEL TOOK 40ms
```

With markers loaded from bytes:
```
UPDATING (97 add, 0 change, 2 delete) MARKERS ON PLATFORM CHANNEL TOOK 475ms
```

So almost 12x slower when markers are loaded from bytes.

I still think the problem is that, in the case of markers loaded from bytes, a copy of the `ByteData` is sent for every markers on update, leading to a huge `update` object being sent to the `platform channel`. 

What do you guys think @chinmaygarde @escamoteur ?

**Comment by danielgomezrico** [comment]
@theochampion @chinmaygarde @escamoteur ping, any updates on this?

**Comment by C4s4r** [comment]
I'm not quite sure but I think I still encounter the same issue.
My app gets a list of local gas stations around the user which varies between 10-500 and creates a custom marker widget (displaying brand icon and price) for each one, which is then transformed into a list of bitmaps by using [this](https://github.com/itsJoKr/markers_generator_sample/blob/master/lib/gist/Gist.dart).

The final markers for the map are then created by using `BitmapDescriptor.fromBytes`.
Performance is really low since using this way of creating dynamic markers for the map, even if the amount of markers is low (<20).

**Comment by Jure-BB** [comment]
@C4s4r I have the same issue. A lot of PNG encoding & decoding can be resource intensive. Sadly, BitmapDescriptor.fromBytes supports only PNG format.

**Comment by maheshmnj** [comment]
Hey @theochampion,
Can you please provide a minimal sample code to reproduce the issue?

Thanks.

**Comment by TahaTesser** [comment]
Without additional information, we are unfortunately not sure how to resolve this issue. 
We are therefore reluctantly going to close this bug for now. 
Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!  
Thanks for your contribution.

Could everyone who still has this problem please file a new issue with the exact description of what happens, logs, and the output of `flutter doctor -v`.
All system setups can be slightly different, so it's always better to open new issues and reference related issues.

**Comment by theochampion** [comment]
@TahaTesser @maheshmnj The example code for this library actually already show this behavior, when increasing the number of rendered markers

**Comment by TahaTesser** [comment]
Hi @theochampion 
Thanks for the ping, I am able to reproduce the issue.

 have a created complete runnable code sample using your code snippet and Indeed the performance takes a hit after generating markers from byes.
 
  It;'s hard to demo it on video but here it is anyway


https://user-images.githubusercontent.com/48603081/125638157-a3e20a8c-f7e7-44e5-b012-fd508a8fe526.mp4


 
 
##  Used the following minimal code sample

<details> 
<summary>minimal code sample</summary> 

```dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';
import 'dart:ui' as ui;

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      showPerformanceOverlay: true,
      title: 'Material App',
      theme: ThemeData.dark(),
      home: Home(),
    );
  }
}

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  Map<MarkerId, Marker> markers = <MarkerId, Marker>{};

  Future<BitmapDescriptor> generatePostClusterMarker(
      String text, double sizeFactor, double pixelRatio) async {
    print(pixelRatio);
    Function wRatio = (input) => input * pixelRatio;
    final stopwatch = Stopwatch()..start();
    final ui.PictureRecorder recorder = ui.PictureRecorder();
    final Canvas canvas = Canvas(recorder);

    final TextSpan textSpan = TextSpan(
      style: TextStyle(
          color: Colors.pink,
          fontSize: wRatio(13 * sizeFactor),
          fontWeight: FontWeight.w900),
      text: text,
    );

    final TextPainter textPainter = TextPainter(
        text: textSpan,
        textDirection: TextDirection.ltr,
        textAlign: TextAlign.center);
    textPainter.layout();

    double diameter = 100 * sizeFactor;
    double blurSigma = wRatio(5);
    Size boxSize = Size(diameter, diameter);
    Size outputSize =
        Size(boxSize.width + 2 * blurSigma, boxSize.height + 2 * blurSigma);

    // paint shadow
    canvas.drawCircle(
        Offset(outputSize.width / 2, outputSize.height / 2),
        diameter / 2,
        Paint()
          ..color = Colors.black.withAlpha(50)
          ..maskFilter = MaskFilter.blur(BlurStyle.normal, blurSigma / 2));
    // paint rounded box
    canvas.drawCircle(
        Offset(outputSize.width / 2, outputSize.height / 2),
        diameter / 2,
        Paint()
          ..color = Colors.white
          ..isAntiAlias = true
          ..style = PaintingStyle.fill
          ..filterQuality = FilterQuality.high);
    // paint text
    textPainter.paint(
        canvas,
        Offset((outputSize.width / 2) - (textPainter.size.width / 2),
            (outputSize.height / 2) - (textPainter.size.height / 2)));

    final ui.Picture picture = recorder.endRecording();

    final ByteData? pngBytes = await picture
        .toImage(outputSize.width.toInt(), outputSize.height.toInt())
        .then((image) => image.toByteData(format: ui.ImageByteFormat.png));
    print(
        'Marker $text [${outputSize.width}x${outputSize.height}] generated in ${stopwatch.elapsed.inMilliseconds}');

    picture.dispose();
    return BitmapDescriptor.fromBytes(pngBytes!.buffer.asUint8List());
  }

  void _add(
      {required String id, required double lat, required double long}) async {
    String markerIdVal = id;
    final MarkerId markerId = MarkerId(markerIdVal);

    final random = Random();

    final BitmapDescriptor bitmapDescriptor =
        await generatePostClusterMarker('Flutter', 2, 2);

    // creating a new MARKER
    final Marker marker = Marker(
      markerId: markerId,
      position: LatLng(lat, long),
      infoWindow: InfoWindow(title: markerIdVal, snippet: '*'),
      icon: bitmapDescriptor,
      onTap: () {
        // _onMarkerTapped(markerId);
      },
    );

    setState(() {
      // adding a new marker to map
      markers[markerId] = marker;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Maps + BitmapDescriptor'),
      ),
      body: GoogleMap(
        initialCameraPosition: const CameraPosition(
          target: LatLng(37.42796133580664, -122.085749655962),
          zoom: 5,
        ),
        markers: Set<Marker>.of(markers.values), // YOUR MARKS IN MAP
      ),
      floatingActionButton: FloatingActionButton(
        child: Icon(Icons.add),
        onPressed: () async {
          for (var i = 0; i < 25; i++) {
            _add(
                id: 'marker_$i',
                lat: 37.427961335806642 + i,
                long: -122.085749655962 + i);

            await Future.delayed(const Duration(milliseconds: 500));
          }
        },
      ),
    );
  }
}

``` 

</details>

## logs

 
|  `stable`  | `master` |
| :---: |:---: |
|  ❌  |   ❌  | 

Check `flutter doctor -v` outputs for each channel below

<details> 
<summary>flutter doctor -v</summary> 
	
```console
[✓] Flutter (Channel stable, 2.2.3, on Microsoft Windows [Version 10.0.22000.65], locale en-US)
    • Flutter version 2.2.3 at C:\Users\Taha\Code\flutter_stable
    • Framework revision f4abaa0735 (13 days ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\Taha\Code\android-sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_SDK_ROOT = C:\Users\Taha\Code\android-sdk
    • Java binary at: C:\Users\Taha\Code\android-studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2019 16.10.3)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.10.31424.327
    • Windows 10 SDK version 10.0.19041.0

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).

[✓] VS Code (version 1.58.1)
    • VS Code at C:\Users\Taha\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.24.0

[✓] Connected device (4 available)
    • Redmi K20 Pro (mobile) • def0ad20 • android-arm64  • Android 10 (API 29)
    • Windows (desktop)      • windows  • windows-x64    • Microsoft Windows [Version 10.0.22000.65]
    • Chrome (web)           • chrome   • web-javascript • Google Chrome 91.0.4472.124
    • Edge (web)             • edge     • web-javascript • Microsoft Edge 91.0.864.67

! Doctor found issues in 1 category.
```

```console
[✓] Flutter (Channel master, 2.4.0-5.0.pre.47, on Microsoft Windows [Version 10.0.22000.65], locale en-US)
    • Flutter version 2.4.0-5.0.pre.47 at C:\Users\Taha\Code\flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision e990f4b912 (9 hours ago), 2021-07-14 01:26:04 -0400
    • Engine revision 093c4ed1a3
    • Dart version 2.14.0 (build 2.14.0-312.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\Taha\Code\android-sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_SDK_ROOT = C:\Users\Taha\Code\android-sdk
    • Java binary at: C:\Users\Taha\Code\android-studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2019 16.10.3)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.10.31424.327
    • Windows 10 SDK version 10.0.19041.0

[✓] Android Studio (version 4.2.0)
    • Android Studio at C:\Users\Taha\Code\android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)

[✓] VS Code (version 1.58.1)
    • VS Code at C:\Users\Taha\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.24.0

[✓] Connected device (5 available)
    • Redmi K20 Pro (mobile)  • def0ad20 • android-arm64   • Android 10 (API 29)
    • Windows (desktop)       • windows  • windows-x64     • Microsoft Windows [Version 10.0.22000.65]
    • Windows (UWP) (desktop) • winuwp   • windows-uwp-x64 •
    • Chrome (web)            • chrome   • web-javascript  • Google Chrome 91.0.4472.124
    • Edge (web)              • edge     • web-javascript  • Microsoft Edge 91.0.864.67

• No issues found!
```
		
</details>

✅ : No Issue ❌: Issue reproduced

**Comment by PiotrMitkowski** [comment]
Hey guys, I came to the same issue while working on my app - on iOS, when loading bitmap descriptors from bytes the overall performance drops when comparing to loading descriptors from assets. I was wondering if it can be solved by adding an extra parameter to `fromBytes` - a string identifier - so it could cache already loaded descriptors on the platform side.

**Comment by ritolika** [comment]
Stumbled upon this issue while trying to implement animated markers where the total frame amount is > 2000 for each animation. I started digging in the source code for the package and native plugins.  As stated earlier the problem with performance is that every time the native platform receives a marker update and has to extract the icon information sent via the serialized bitmap descriptor it reads the whole byte array and also creates a new instance of UIImage. Resulting in very slow performance when switching byte array based icons on markers rapidly or often. What I did to solve this issue was to fork the current google_maps_flutter project and implement some changes. 

In GoogleMapController I added a method `setCachedBitmaps(Map<int, Uint8List> bitmapByIndex)`. The method takes a map of entries where the key is a unique id that goes along with a png byte array value. So that index -> pngByteArray. Then I added a fourth way of initializing a BitmapDescriptor by adding `static BitmapDescriptor fromCacheIndex(int cacheIndex)` and as you already guessed cacheIndex corresponds to the index in bitmapByIndex. 

Before going further I'd like to note a few things. I'm not experienced in Objective-C thus the code may be optimized further. And also as setCachedBitmaps is connected to the instance of MapController the cached bitmaps are only available for the Map which has the MapController of which setCachedBitmaps was called on. Also, this can be further improved by natively changing so that the caching happens somewhere where it's independent of instances but also accessible by every controller instance. 

How to use currently:
Replace the current google_maps_flutter with the public repo hosted by me (click in the profile).
how your pubspec.yml should look:
`dependencies:
  flutter:
    sdk: flutter
  google_maps_caching_flutter:
    git:
      url: git://github.com/ritolika/flutter_google_maps_caching.git
      path: google_maps_caching_flutter`
When creating a GoogleMap widget you should wrap the MapController with a Completer to set your bitmaps like this:
`GoogleMapController controller = await _controller.future;
      await controller.setCachedBitmaps(component3d.cacheData);`
Or use Future.then to set accordingly. 
When you've made sure that the cachedBitmaps are set you can initialize a BitmapDescriptor with `BitmapDescriptor.fromCacheIndex(int cacheIndex)`. 
The iOS platform will the take the cached UIImage instead of loading a new one each time. Resulting in a great performance which makes stuff like the video possible (run on iPhone 7 with Flutter ofc).
I will recode the Objective-C part so that the caching is done globally so that it can be used by any map instance at any time. 
After that, I will write the same implementation for Android. Then I'll post a more extensive guide as to how to set it up.


https://user-images.githubusercontent.com/9406612/132355765-7d8f3f0d-29a3-4830-8783-03728a11a816.mp4

**Comment by manishrelani** [comment]
Appreciate your response to my query. Yes, your approach to this totally makes sense. 
Only I could not comprehend this : 
"You could also calculate the actual time it took from the last frame update as the timer not always executes with the same delay every time".
It would be great if you could explain this. 

PS. I deleted my comment after I posted, thinking that would be off-topic to this thread.

**Comment by ritolika** [comment]
Yes, I see why you removed the comment you might be right about it being
off-topic but I've also seen multiple comments not providing anything to
the conversation.

What I meant is that when you have a repeating task you can never be 100%
sure that the task is being run with the same delay each time. For example,
if you target an update every 20ms it might be stalled or delayed further
by an app freeze or similar and the actual time between the last update
might, for example, go up to 35ms. If every time the timer runs check the
time it took since the last update you'll find out the actual delay. With
that information, you can choose to skip a frame for example if the delay
doubled. That's what I meant. I can clarify further if necessary.

/Alan

**Comment by AristideVB** [comment]
Any updates on this ? @ritolika I would love to contribute in opening a PR

**Comment by Fantome5** [comment]
Any updates on this? Experiencing poor performance rendering ~200 markers.

**Comment by AristideVB** [comment]
@Fantome5 Please put a thumbs up (👍) to the original post (and ask your colleagues to do the same :) ). It will help to make the issue visible to the Flutter team.
